src/tfs/tfs.c

```diff
@@ -439,22 +488,99 @@ static extent create_extent(fsfile f, range r, merge m)
     extent ex = allocate_extent(h, r, block_start, alloc_bytes);
     if (ex == INVALID_ADDRESS)
         halt("out of memory\n");
-    assert(rangemap_insert(f->extentmap, &ex->node));
+    ex->uninited = uninited;
+
+    return ex;
+}
+
+static void destroy_extent(filesystem fs, extent ex)
+{
+    deallocate_u64((heap)fs->storage, ex->block_start, ex->allocated);
+    deallocate(fs->h, ex, sizeof(*ex));
+}
+
+static void add_extent_to_file(fsfile f, extent ex, merge m)
+{
+    heap h = f->fs->h;
 
     // XXX encode this as an immediate bitstring
-    tuple e = timm("length", "%ld", length);
-    string offset = aprintf(h, "%ld", block_start);
+    tuple e = timm("length", "%ld", range_span(ex->node.r));
+    string offset = aprintf(h, "%ld", ex->block_start);
     table_set(e, sym(offset), offset);
-    string allocated = aprintf(h, "%ld", alloc_bytes);
+    string allocated = aprintf(h, "%ld", ex->allocated);
     table_set(e, sym(allocated), allocated);
-    symbol offs = intern_u64(r.start);
+    if (ex->uninited) {
+        table_set(e, sym(uninited), allocate_buffer(h, 0));
```

Comment:
It's just a placeholder value. For making a constant value, do you mean how filesytem_creat() does to set the length of a newly created file to 0? I.e. declaring a static buffer and initializing it with wrap_buffer_cstring() on first use? I guess for the "uninited" symbol we could just pass an empty string to wrap_buffer_cstring()?

Reply:
Sure that could work - or even just do it as you're doing it now, but stash it indefinitely and reuse it rather than allocate a new buffer each time. Beyond the scope of this PR, I was making more of a general comment about how tuples are used for flags (i.e. when the value doesn't really matter). Maybe we could create a global "null" value (buffer) in init_tuples() using the wrap cstring thing (at least that way it's somewhat enforced as a constant - a write to a constant string should cause a fault, right?)...

Reply:
OK, creating a global "null" value seems the best thing to do, if one tries to write to it using the buffer API, it will cause a fault since it's a wrapped buffer.
I'm going to declare a `value null_value(void)` function in tuple.h, which returns the pointer to such global value defined (and initialized) in tuple.c.